x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/DelimitedFileStructureFinder.java

```diff
@@ -529,16 +536,25 @@ static boolean canCreateFromSample(List<String> explanation, String sample, int
                         --fieldsInThisRow;
                     }
 
-                    if (fieldsInLastRow != fieldsInFirstRow) {
-                        explanation.add("Not " + formatName + " because row [" + (numberOfRows - 1) +
-                            "] has a different number of fields to the first row: [" + fieldsInFirstRow + "] and [" +
-                            fieldsInLastRow + "]");
-                        return false;
+                    if (fieldsInThisRow != fieldsInFirstRow) {
+                        illFormattedRows.add(fieldsInFirstRow);
+                        continue;
                     }
 
                     fieldsInLastRow = fieldsInThisRow;
                 }
 
+                // We should only allow a certain percentage of ill formatted rows
+                // as it may effects our sample sizing and down stream processing
+                if (illFormattedRows.size() > (allowedFractionOfBadLines * numberOfRows) ) {
```

Comment:
I think it should round up so that if the number of rows is small there's still a tolerance

```suggestion
                if (illFormattedRows.size() > Math.ceil(allowedFractionOfBadLines * numberOfRows) ) {
```

But also, this needs to short-circuit.  Imagine we have been told to analyse 50000 lines that are _not_ delimited.  With this change the code is now going to analyse all 50000 for all 4 default delimiters before ruling out the delimited formats.  And then it's going to provide 4 explanations each listing 49999 line numbers.  Therefore I think that:

1. This test needs to be inside the loop above, and `break` when the condition fails
2. The explanation message that lists bad rows needs to be rephrased to take this into account (see my other comment)